lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+@six.add_metaclass(abc.ABCMeta)
+class GitRef(object):
+    """A wrapper that knows how to format git refspecs in multiple contexts."""
+    known_types = ['Branch', 'Commit', 'Tag']  # type: ClassVar[List[str]]
+    ref_type = None                            # type: ClassVar[str]
+    _ref = None                                # type: str
+
+    Tag = None                # type: ClassVar[Type[GitTag]]
+    Commit = None             # type: ClassVar[Type[GitCommit]]
+    Branch = None             # type: ClassVar[Type[GitBranch]]
+
+    def __init__(self, ref):
+        # type: (str) -> None
+        if self.ref_type not in self.known_types:
```

Comment:
Does capitalization matter here? E.g., can we standardize so anything with the string will work (e.g., so here do lower() and then capitalize() ?

Reply:
Oh well this part was just intended to be an especially obnoxious logical check -- when I say in the comment just below "it cannot arise from invalid user input" I mean it's not going to happen. It's more intended to be a reminder that GitRef should never be instantiated by itself, only through one of those three subclasses, known in advance.

It's possible this check could be removed, since it's really unimportant to the overall logic.